Convert to custom table

ethitter opened the pull request:
By introducing a custom table with columns and indices specific to the plugin's needs, issues with duplicate events, race conditions that create excessive events, and stale caches are all alleviated. Additionally, by adding a number of utility functions, the plugin's internal consistency is greatly improved.

Some of the changes herein:
* table is created, version stored in an option, mirroring Core's behaviour
* table includes a unique constraint to prevent duplicate entries from being added
* CLI to purge legacy CPT data is introduced
* various classes no longer contain methods to retrieve events thanks to the introduction of flexible helper functions

Outstanding concerns:
* table lacks sufficient indices; currently, they're mostly used to block duplicates, not for performance
* table is created on first load, which could lead to a race condition of table creation; I'm not sure it's safe to restrict to `is_admin()`, though, as WP will keep trying to create entries

Fixes #24 
Fixes #39 
Fixes #63 
Fixes #64 
See #84
Fixes #85 
Fixes #90 
Fixes #91
Fixes #94

ethitter commented:
Optimistically, duplicates will be less of an issue, obviating #49 and #52.

ethitter commented:
I'd still like to address #37, but don't think it's a blocker for this PR.

ethitter commented:
The other concern I have is around the performance of the `job_exists()` method in `Events_Store`. Some caching might be warranted.

joshbetz commented:
Noting the schema here:

```sql
CREATE TABLE `{$this->get_table_name()}` (
	`ID` bigint(20) unsigned NOT NULL AUTO_INCREMENT,

	`timestamp` bigint(20) unsigned NOT NULL,
	`action` varchar(255) NOT NULL,
	`action_hashed` varchar(32) NOT NULL,
	`instance` varchar(32) NOT NULL,

	`args` longtext NOT NULL,
	`schedule` varchar(255) DEFAULT NULL,
	`interval` int unsigned DEFAULT 0,
	`status` varchar(32) NOT NULL DEFAULT 'pending',

	`created` datetime NOT NULL,
	`last_modified` datetime NOT NULL,

	PRIMARY KEY (`ID`),
	UNIQUE KEY `ts_action_instance` (`timestamp`, `action` (191), `instance`, `status`)
) ENGINE=InnoDB;
```

joshbetz commented:
> table is created on first load, which could lead to a race condition of table creation; I'm not sure it's safe to restrict to is_admin(), though, as WP will keep trying to create entries

How hard would it be to maintain back-compat for the custom post type and force the migration to a custom table to happen via wp-cli?

joshbetz commented on includes/wp-cli/class-events.php:
```diff
@@ -211,64 +222,93 @@ private function format_events( $events ) {
+			return 1;
+		}
+
+		// If timestamps are equal, consider action
```
`return strnatcasecmp($first['action'], $second['action'])`?

http://php.net/manual/en/function.strnatcasecmp.php

ethitter replied:
:man_facepalming:

joshbetz commented on includes/wp-cli/class-events.php:
```diff
@@ -211,64 +222,93 @@ private function format_events( $events ) {
+
+		if ( $first_timestamp > $second_timestamp ) {
+			return 1;
+		}
```
Could simplify these two conditionals to

```
if ( $first_timestamp !== $second_timestamp ) {
   return $first_timestamp - $second_timetamp;
}
```

joshbetz commented on includes/class-events-store.php:
```diff
@@ -0,0 +1,592 @@
+
+		$updates = array(
+			'status'   => self::STATUS_COMPLETED,
+			'instance' => wp_rand( 1000000, 999999999 ), // Breaks unique constraint, and can be recreated from entry's remaining data
```
Why do we need to break the unique constraint here?

ethitter replied:
With non-recurring events scheduled for specific times, like the ones that publish scheduled posts, if the event is cleared and rescheduled, the rescheduled event can't be marked as completed until the plugin's cleanup job runs. As a result, the rescheduled event would be fired repeatedly, for up to an hour. Since instance is a hash of other data we have, overwriting it breaks the unique without destroying data.

Events aren't deleted directly for performance reasons. That's true in the CPT version, which carried over with this PR. Completed events will be cleared less-frequently after this PR--every hour instead of every ten minutes--to take advantage of indices and reduce disruptions to InnoDB's clustering.

Looking at this now, I'm not sure that `wp_rand()` is really necessary; `mt_rand()` is probably sufficient.

ethitter commented on includes/class-events-store.php:
```diff
@@ -0,0 +1,592 @@
+		}
+
+		// Do not sort, otherwise index isn't used
+		$jobs = $wpdb->get_results( $wpdb->prepare( "SELECT * FROM {$this->get_table_name()} WHERE status = %s LIMIT %d,%d;", $args['status'], $offset, $args['quantity'] ), 'OBJECT' );
```
Ah, bad documentation. It doesn't use an index (right now), but adding a sort changes it from a simple where query to one needing a filesort.

ethitter replied:
Forgot to note that there is an index on status, care of the unique. Adding an index on status alone doesn't help, unless this query is split into one that retrieves IDs, and another that retrieves those rows using `IN`. Since it's a simple where, I'm not sure splitting is worth it.

joshbetz replied:
Does that index get used here? I think a composite index will only get used if you're querying a prefix of the index. i.e. we could use the unique key to search (timestamp), (timestamp, action), etc., but not status by itself.

https://dev.mysql.com/doc/refman/5.7/en/multiple-column-indexes.html

> If the table has a multiple-column index, any leftmost prefix of the index can be used by the optimizer to look up rows. For example, if you have a three-column index on (col1, col2, col3), you have indexed search capabilities on (col1), (col1, col2), and (col1, col2, col3).

ethitter replied:
You're right, the composite index doesn't get used. In normal circumstances, there aren't enough rows to use an index on `status`, but I think it's worth adding for safety's sake.

ethitter commented:
> How hard would it be to maintain back-compat for the custom post type and force the migration to a custom table to happen via wp-cli?

It wouldn't be simple. We'd need an abstraction layer to switch various lookups from one source to the other, which would be complicated by the different data format that the custom table affords.

To help with the initial table creation, the function now locks itself using `wp_cache_add()`, so the issue is less of a concern to me than when I opened the PR (2f019c6).

ethitter commented on includes/class-events-store.php:
```diff
@@ -0,0 +1,592 @@
+	 *
+	 * Uses a direct query to avoid stale caches that result in duplicate events
+	 */
+	public function job_exists( $timestamp, $action, $instance, $return_id = false ) {
```
This is one of several vestigial patterns in this plugin, which I'm not thrilled about. I'll look at fixing this.

ethitter replied:
Thanks for calling this out! It turns out that `job_exists()` isn't really needed at all, just the ID-returning version of it. :man_facepalming:

ethitter commented:
> I agree we need one or two indexes -- probably `status` and maybe `action` or `action_hashed`.

I was mistaken. 😞 

`status` and `action` are part of the unique index already in place. I added a standalone index for each of the three columns, but it proved unnecessary. Explaining each of the queries the plugin uses showed that the `ts_action_instance` index was used instead of a column-specific one, even when `action_hashed` was in the query.

ethitter commented:
With the exception of the weird pattern that is `job_exists()` returning an ID on demand, I've addressed all of Josh's feedback. 

More feedback welcome!!!

And yes, I'm looking at making `job_exists()` less...strange. We're living with some poor decisions I made early on in this plugin's development.

mjangda commented on includes/class-events-store.php:
```diff
@@ -0,0 +1,592 @@
+		$quantity = 100;
+
+		do {
+			$jobs_posts = $this->get_jobs( array(
```
Minor: rename to `$jobs` since they're not posts anymore :)

mjangda commented on includes/class-events-store.php:
```diff
@@ -0,0 +1,592 @@
+	 */
+	public function block_creation_if_job_exists( $job ) {
+		$instance = md5( maybe_serialize( $job->args ) );
+		if ( $this->job_exists( $job->timestamp, $job->hook, $instance ) ) {
```
Minor: move out function call from `if`.

mjangda replied:
Alternatively, following Josh's suggestion, if the method returns a `bool`, we could just do `return $this->job_exists( ... )`

mjangda commented on includes/class-events-store.php:
```diff
@@ -0,0 +1,592 @@
+		if ( is_array( $jobs ) ) {
+			$jobs = array_map( array( $this, 'format_job' ), $jobs );
+		} else {
+			$jobs = false;
```
Can we just return an empty array?

mjangda commented on includes/class-events-store.php:
```diff
@@ -0,0 +1,592 @@
+		$query_replacements = array();
+
+		if ( isset( $attrs['ID'] ) ) {
+			$query                = "SELECT * FROM {$this->get_table_name()} WHERE ID = %d";
```
This should probably be its own function (`get_job_by_id`)?

ethitter replied:
My intent was to make this method a bit like `WP_Query` in its flexibility. The use case for looking a job up by ID is limited to the plugin's CLI commands, so I can move this bit somewhere specific to that context.

mjangda commented on includes/class-events-store.php:
```diff
@@ -0,0 +1,592 @@
+
+		$query .= ' LIMIT 1';
+
+		$query = $wpdb->prepare( $query, $query_replacements );
```
The `query_replacements` approach seems a bit fragile. I think we should `prepare` in context (i.e. whenever we build each SQL statement).

If anything, `$query` should be an array of prepared SQL statements which is `implode`-ed with a space.

ethitter replied:
It's...different, that's for sure. Again, this comes from spending too much time staring at Core. With the ID-lookup bit going away, this approach won't really be needed anymore.

mjangda commented on includes/class-events-store.php:
```diff
@@ -0,0 +1,592 @@
+	 *
+	 * @return array
+	 */
+	private function find_cron_array_differences( $new, $old ) {
```
Minor: technically new and old are not accurate names since we swap the order to find deletes.

mjangda commented on includes/class-events-store.php:
```diff
@@ -0,0 +1,592 @@
+		global $wpdb;
+
+		// Skip count if already performed
+		if ( $count_first ) {
```
What do we need the `count_first` for?

ethitter replied:
When it's called from CLI, the count query is already run, so I wanted to avoid running it again or blocking the delete call. Now that I've described it, it should be `$force` instead.

mjangda commented on includes/class-internal-events.php:
```diff
@@ -113,17 +113,30 @@ public function is_internal_event( $action ) {
+		$page     = 1;
+		$quantity = 25;
+
+		do {
```
Do we really need this loop here? The missed schedule event runs every minute so the next batch of 25 would just be published in the next ping? (Not to mention that something must be seriously wrong if we have more than 25 missed schedule posts.)

mjangda commented on includes/class-internal-events.php:
```diff
@@ -132,30 +145,47 @@ public function force_publish_missed_schedules() {
+		$page     = 1;
+		$quantity = 25;
+
+		do {
```
Same as above; the loop is probably not necessary. If we expect lots of posts, can we just bump the limit to 100 or something similar (especially since we're not fetching a lot of data)?

ethitter replied:
I'll bump the limit to 100, but I want to keep this loop. We have a few clients who schedule weeks worth of posts, and get close to the cap of 100. I'd rather confirm everything than risk missing something. This only runs every ten minutes, so I'm not that concerned about the load it'll introduce or the time it'll take to process.

mjangda commented on includes/class-internal-events.php:
```diff
@@ -132,30 +145,47 @@ public function force_publish_missed_schedules() {
-				$gmt_time        = strtotime( get_gmt_from_date( $future_post->post_date ) . ' GMT' );
-				$timestamp       = wp_next_scheduled( 'publish_future_post', array( $future_post->ID ) );
+			if ( empty( $future_posts ) ) {
+				break;
```
This early `break` breaks the pattern used in other parts of the codebase.

ethitter replied:
Yeah, I have no idea...

ethitter commented on includes/class-events-store.php:
```diff
@@ -0,0 +1,592 @@
+	/**
+	 * Create the plugin's DB table when necessary
+	 */
+	protected function prepare_tables() {
```
Initially I did, but until the table is created, every request tries to schedule the events a site defined. The `wp_cache_add()` is meant to guard against this. Instead, or perhaps in addition, the option should be set sooner, to cut down on how many processes can create the table.

ethitter commented:
I want to look at #94 as part of this. Two sites created nearly 4,000 entries apiece, likely thanks to that.

ethitter commented on includes/class-events-store.php:
```diff
@@ -0,0 +1,592 @@
+		global $wpdb;
+
+		// Define schema and create the table
+		$schema = "CREATE TABLE `{$this->get_table_name()}` (
```
I'd _really_ rather not. I'm familiar with the FG and related P2 discussion, and the arguments made there don't apply to this table, at least as I understand them. There's no use case for joining the plugin's table against any others--the IDs in `a8c_cron_control_jobs` don't correspond to any ID in any other table. If we found ourselves in a situation where a join was necessary, I'd rather we alter the table (and add appropriate indexes) than introduce a dependency on another table.

I'm also opposed to this change because I don't want to repeat the full column name throughout the codebase, nor do I really want to introduce a constant, helper, etc. to avoid repeating `a8c_cton_control_jobs_id`.

I understand the reasoning for tables on wpcom, or that might be joined against others, but don't feel it adds any value in this plugin.

ethitter commented on includes/class-events-store.php:
```diff
@@ -0,0 +1,592 @@
+		if ( 1 === $table_count ) {
+			update_option( self::DB_VERSION_OPTION, self::DB_VERSION, true );
+		} else {
+			delete_option( self::DB_VERSION_OPTION );
```
Yeah, I've been unsure of what to do here, because if the table doesn't exist, cron is completely broken as long as the plugin is active. If the table isn't created, the plugin shouldn't override `get|update_option()` calls. Setting an option or cache key to disable the filters in `class_init()` might be worthwhile, as well as setting the option to the expected version as a short-circuit to future upgrades. An action, which we can hook separately from mu-plugins, could handle alerting.

ethitter commented on includes/class-events-store.php:
```diff
@@ -0,0 +1,592 @@
+	 * Build array of valid event statuses
+	 */
+	public function get_allowed_statuses() {
+		return array(
```
I think so. There was a time when I thought this might be dynamic depending on how/where it was called, but that hasn't proven necessary.

ethitter commented:
@mjangda & @joshbetz - this is ready for another look. There's an open question around stampedes with the initial table creation, but I've addressed the remaining feedback and left a few thoughts on how to deal with the table.

mjangda commented on tests/bootstrap.php:
```diff
@@ -20,6 +20,15 @@ function _manually_load_plugin() {
+	remove_filter( 'schedule_event', '__return_false' );
+	add_filter( 'pre_option_cron', array( \Automattic\WP\Cron_Control\Events_Store::instance(), 'get_option' ) );
+	add_filter( 'pre_update_option_cron', array( \Automattic\WP\Cron_Control\Events_Store::instance(), 'update_option' ), 10, 2 );
+	add_filter( 'schedule_event', array( \Automattic\WP\Cron_Control\Events_Store::instance(), 'block_creation_if_job_exists' ) );
```
Minor: if we add `use \Automattic\WP\Cron_Control\Events_Store` to the top of this file, we can simplify the calls to this namespace:

```
add_filter( 'pre_update_option_cron', array( Events_Store::instance(), 'update_option' ), 10, 2 );
add_filter( 'schedule_event', array( Events_Store::instance(), 'block_creation_if_job_exists' ) );
```

ethitter replied:
Good point. I honestly don't know why I didn't use `use` anywhere in the tests or CLI classes.
